sympy/polys/numberfields/modules.py

```diff
@@ -0,0 +1,1794 @@
+        return self._QQ_col
+
+    def to_parent(self):
+        """
+        Transform into a :py:class:`~.ModuleElement` belonging to the parent of
+        this element's module.
+        """
+        if not isinstance(self.module, Submodule):
+            raise ValueError('Not an element of a Submodule.')
+        return make_mod_elt(
+            self.module.parent, self.module.matrix * self.col,
+            denom=self.module.denom * self.denom)
+
+    def to_ancestor(self, anc):
+        """
+        Transform into a :py:class:`~.ModuleElement` belonging to a given
+        ancestor of this element's module.
+
+        Parameters
+        ==========
+
+        anc : :py:class:`~.Module`
+
+        """
+        if anc == self.module:
+            return self
+        else:
+            return self.to_parent().to_ancestor(anc)
+
+    def over_power_basis(self):
+        """
+        Transform into a :py:class:`~.PowerBasisElement` over our
+        :py:class:`~.PowerBasis` ancestor.
+        """
+        e = self
+        while not isinstance(e.module, PowerBasis):
+            e = e.to_parent()
+        return e
+
+    def is_compat(self, other):
+        """
+        Test whether other is another :py:class:`~.ModuleElement` with same
+        module.
+        """
+        return isinstance(other, ModuleElement) and other.module == self.module
+
+    def make_compat(self, other):
+        """
+        Try to make a compatible pair of :py:class:`~.ModuleElement`, one
+        equivalent to this one, and one equivalent to the other.
+
+        Explanation
+        ===========
+
+        We search for the nearest common ancestor module for the pair of
+        elements, and represent each one there.
+
+        Returns
+        =======
+
+        Pair ``(e1, e2)``
+            Either each of ``e1, e2`` is a :py:class:`~.ModuleElement` and they
+            belong to the same :py:class:`~.Module`, or else both are ``None``,
+            if no common ancestor could be found.
+
+        """
+        if self.module == other.module:
+            return self, other
+        nca = self.module.nearest_common_ancestor(other.module)
+        if nca is not None:
+            return self.to_ancestor(nca), other.to_ancestor(nca)
+        return None, None
+
+    def __eq__(self, other):
+        if self.is_compat(other):
+            return self.QQ_col == other.QQ_col
+        return NotImplemented
+
+    def equiv(self, other):
+        """
+        A :py:class:`~.ModuleElement` may test as equivalent to a rational
+        number or another :py:class:`~.ModuleElement`, if they represent the
+        same algebraic number.
+
+        Explanation
+        ===========
+
+        When *other* is another :py:class:`~.ModuleElement`, the test for
+        equivalence requires not just that the two elements represent the same
+        algebraic number, but that they have the same :py:class:`~.PowerBasis`
+        ancestor.
```

Comment:
It seems odd to me that it's okay for other to be something completely different that isn't even a ModuleElement but if it is a ModuleElement then it needs to have exactly the same ancester. Why is it useful to have a method that works like this?

Reply:
Well, it was restricted to these cases only because they are the easy ones (and the only ones I needed!).

Without a common `PowerBasis` ancestor, it essentially means asking whether an element of `QQ(alpha)` and an element of `QQ(beta)` are equal, with `alpha` and `beta` being algebraic numbers not necessarily bearing any obvious relation to one another.

I guess it wouldn't hurt to handle this case too, using the `to_number_field` function. It just means bringing in some heavy machinery that's not required for the cases currently handled.

Reply:
On second thought, I think the remaining case (`ModuleElement`s lacking a common ancestor) has to either wait, or just be dropped. It's really not needed at all, and it was just going to be a "bonus" for "completeness".

I think it's perfectly fine to raise `UnificationFailed` here. Also I now make it clear in the docstring that this is a deliberate limitation.

The issue is, I am actually having trouble getting `minpoly.py` and `subfield.py` to answer this question, in the form in which it's given. I'm not yet able to say whether they actually _can't_ do it, or I simply haven't figured out how. In any case, it doesn't seem to be easy.

If it turns out to be easy, I will come back and supply this case. If it turns out to be hard, then we might be talking about some sizable new developments in `minpoly.py` and/or `subfield.py`, which would have to be done in a  separate PR.